Dedup Backend Initial Implementation

ikreymer opened the pull request:
Fixes #2867 

Set to feature branch as base.

The backend implementation involves:
- A new CollIndex CRD type
- Operator that manages the new CRD type, creating a new Redis instance when the index should exist
- Operator starts the crawler in 'indexer' mode (will be available from https://github.com/webrecorder/browsertrix-crawler/issues/884)
- Collection has a new `hasDedupeIndex` field
- Workflows have a new 'dedupeCollId' field for dedup while crawling. The `dedupeCollId` must also be a collection that the crawl is auto-added to.
- There is a new waiting state: `waiting_for_dedupe_index` that is entered if a crawl is starting, but index is not yet ready.
- Each crawl has bi-directional links for crawls that it requires for dedupe via `requiresCrawls` and other crawls for which this crawl is required via `requiredByCrawls`.
- autoAddCollections automatically updated to always include `dedupeCollId` collection.

For indexing, dependent on version of crawler (1.9.0 beta 0 or higher) that supports indexing mode.

### Testing
This is ready for initial frontend work and testing:
- the `dedupeCollId` can be set on the workflow to enable dedup for future crawls.
- the collection has a `hasDedupIndex` to indicate if an index is enabled for it.
- `kubectl get pods -n crawlers collindex` should work

ikreymer committed:
operator crds:
- add CollIndex crd
- add new operator
- added as btrix-crds 0.2.0

operator collindex:
- redis_dedupe_memory and redis_dedupe_storage configurable
- dedupe_importer_channel can configure crawler channel for index imports

operator crawl:
- add 'waiting_for_dedupe_index' state to indicate crawl is awaiting dedup index
- on crawl success, get list of required crawls deduped against and set 'requiredCrawls' and 'requiredByCrawls'

collection:
- supports creating, updating, deleting CollIndex via CrawlManager
- tracks via hasDedupeIndex
- mark CollIndex object for updates by updating collItemsUpdatedAt on items add or remove
- CollIndex deleted on collection deleted

workflow:
- has dedupeCollId if dedupe is enabled
- dedupeCollId included in autoAdd collections
- update empty string dedupeCollId to remove dedupe collection from workflow

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -146,6 +158,9 @@ async def add_collection(self, oid: UUID, coll_in: CollIn):
             await self.clear_org_previous_slugs_matching_slug(slug, org)
+            if coll_in.hasDedupIndex:
+                await self.toggle_dedup_index(coll.id, True)
+                coll.hasDedupIndex = True
```
Looks like this is getting set after the collection is inserted, not sure it'll actually persist in the db here

ikreymer replied:
Ah, this is intentional, as toggle_dedup_index() checks that its false before setting it to true, so if it was set previously, the toggle would not do anything. Maybe should add a comment about this..

ikreymer replied:
Ended up refactoring this, as it was a bit confusing:
- in collections, explicitly calling create_coll_index() on creation
- calling create_coll_index() on update, or delete_coll_index() when deleting (may need more checks)
- for workflows, calling enable_dedup_index() to enable it if it doesn't exist only
see: https://github.com/webrecorder/browsertrix/pull/2868/commits/ece2aa3350332ef252c2bde5ef6c538869c5f89b

tw4l commented on backend/btrixcloud/colls.py:
```diff
@@ -783,13 +850,16 @@ async def update_collection_dates(self, coll_id: UUID, oid: UUID):
+        # accessing directly to handle both crawls and uploads
         crawl = await self.crawls.find_one({"_id": crawl_id})
-        crawl_coll_ids = crawl.get("collectionIds")
+        crawl_coll_ids = crawl.get("collectionIds") or []
```
```suggestion
        crawl_coll_ids = crawl.get("collectionIds", [])
```
Just a bit more idiomatic

ikreymer replied:
This is just to be safe in case `crawl = {"collectionIds": None}` in which case `crawl.get("collectionIds", [])` will still return `None` but `crawl.get("collectionIds") or []` will return `[]`.
pydantic takes care of this when we use the model since its declared as `collectionIds: Optional[List[UUID]] = []` so will be assigned `[]`, but it might be None in the db, so this is safer for raw access.

tw4l commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -573,6 +587,20 @@ async def update_crawl_config(
+            ):
+                raise HTTPException(
+                    status_code=400, detail="dedup_coll_id_not_in_autoadd"
+                )
```
This logic needs to account for the possibility that the collection referenced by `dedupCollId` may already be an auto-add collection in the original crawl config, prior to the update, which would also be fine so long as it's not being unset in the update. Can look to seed file validation logic just above for an example, and since it's a bit tricky to get right we should have good tests around `dedupCollId` (setting in initial collection add, updating when auto-add collection is already set, updating both together).

tw4l commented on backend/btrixcloud/crawlmanager.py:
```diff
@@ -468,6 +472,31 @@ async def delete_crawl_config_by_id(self, cid: str) -> None:
+        params = {
+            "id": str(collection.id),
+            "oid": str(collection.oid),
+            "collItemsUpdatedAddAt": date_to_str(collection.modified or dt_now()),
```
```suggestion
            "collItemsUpdatedAt": date_to_str(collection.modified or dt_now()),
```
Typo

tw4l commented on backend/btrixcloud/k8sapi.py:
```diff
@@ -76,6 +76,23 @@ async def get_redis_client(self, redis_url):
+        except:
+            if redis:
+                await redis.close()
+
+            return None
```
```suggestion
```

Seems unnecessary

ikreymer replied:
yes, but mypy complains of missing return otherwise.

tw4l commented on backend/btrixcloud/k8sapi.py:
```diff
@@ -274,8 +290,14 @@ async def get_profile_browser(self, browserid):
 
     async def _patch_job(self, crawl_id, body, pluraltype="crawljobs") -> dict:
+        """patch crawl/profile job"""
+        name = f"{pluraltype[:-1]}-{crawl_id}"
```
```suggestion
    async def _patch_job(self, obj_id, body, pluraltype="crawljobs") -> dict:
        """patch crawl/profile job"""
        name = f"{pluraltype[:-1]}-{obj_id}"
```
To help keep it a bit more generic

tw4l commented on backend/btrixcloud/models.py:
```diff
@@ -583,6 +588,8 @@ class CrawlConfigDefaults(BaseModel):
 
     customBehaviors: List[str] = []
 
+    dedupCollId: Optional[UUID] = None
```
If we're adding this to crawl defaults, might be good to add auto-add collections as well? Unless the idea is to just handle the auto-add portion in the workflow frontend?

ikreymer replied:
Yes, the frontend will ensure that dedup coll is also part of auto-add. Not sure if we want auto-add collections as part of the defaults..

tw4l commented on backend/btrixcloud/operator/collindexes.py:
```diff
@@ -0,0 +1,176 @@
+        }
+
+    def get_import_ts(self, spec: CollIndexSpec, status: CollIndexStatus):
+        """returnt rue if a reimport is needed based on last import date"""
```
```suggestion
        """return true if a reimport is needed based on last import date"""
```

tw4l commented on backend/btrixcloud/operator/collindexes.py:
```diff
@@ -0,0 +1,176 @@
+
+    state: TYPE_INDEX_STATES = "initing"
+
+    lastCollUpdated: str = ""
```
Might suggest renaming to `collLastUpdated`, for a sec I thought we were checking other collections

ikreymer replied:
Updated!

tw4l commented on backend/btrixcloud/operator/crawls.py:
```diff
@@ -784,9 +823,25 @@ async def can_start_new(
+            return True
+
+        for index in data.related[COLLINDEX].values():
+            if index.get("status", {}).get("state") == "ready":
```
Is this the opposite of what we want? I think if the coll index is ready, we want to return `False` (i.e. don't need to wait, it's ready), right?

ikreymer replied:
Yes, this was indeed doing the opposite, good catch, should be fixed!

tw4l commented on backend/btrixcloud/uploads.py:
```diff
@@ -163,6 +163,8 @@ async def _create_upload(
                 raise HTTPException(status_code=400, detail="invalid_collection_id")
 
+        print("UUIDS", collection_uuids)
+
```
```suggestion
```

Seems leftover from dev debugging

tw4l reviewed (commented):
A promising start!

I left comments/suggestions throughout where I noticed things. Should also add tests for setting `dedupCollId` for collection and crawlconfig add and update, I don't think it's quite right as-is (at least for collections).

SuaYoo reviewed (commented):
The collection PATCH endpoint will need to accept an empty string for `dedupCollId` so that it can be unset. Currently, the API returns an error that details `Input should be a valid UUID, invalid length: expected length 32 for simple format, found 0`

SuaYoo commented:
Also, updating `dedupCollId` should update `autoAddCollections` on the backend.

SuaYoo commented on backend/btrixcloud/colls.py:
```diff
@@ -783,13 +858,16 @@ async def update_collection_dates(self, coll_id: UUID, oid: UUID):
             await self.update_collection_counts_and_tags(coll_id)
-            await self.update_collection_dates(coll_id, oid)
+            await self.update_collection_dates(
+                coll_id, oid, crawl.get("dedupCollId") != coll_id
```
Can we standardize `dedup` to `dedupe`? I believe this is the more common shorthand (https://github.com/webrecorder/browsertrix/pull/2893#discussion_r2427030781)

ikreymer replied:
Done!

ikreymer committed:
Update backend/btrixcloud/operator/collindexes.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
lastCollUpdated -> collLastUpdated

tw4l commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -603,6 +615,19 @@ async def update_crawl_config(
+                update.autoAddCollections = []
+
+            if update.dedupeCollId not in update.autoAddCollections:
+                update.autoAddCollections.append(update.dedupeCollId)
```
This has two issues:

1. Setting `update.autoAddCollections` to [] has the potential to wipe out any existing auto-add collections from before the update (if a user isn't trying to update the auto-add collections but just add `dedupeCollId`, the previously set auto-add collections would be functionally removed by this update).
2. Relatedly, the update logic needs to check if the `dedupeCollId` is either a) set in the update (currently handled) or b) already set in workflow prior to the update and the update isn't removing it

ikreymer replied:
Good catch! I think this should address both: https://github.com/webrecorder/browsertrix/pull/2868/commits/4fbf4a55e373a92942d83dbea493f84ae7519716

ikreymer committed:
PR review fix:
- if autoAddCollections list exists, add to existing list
- if it does not, just do $addToSet to avoid clearing any existing collections
- consider previous dedupCollId as well and use that if no update

ikreymer commented:
> The collection PATCH endpoint will need to accept an empty string for `dedupCollId` so that it can be unset. Currently, the API returns an error that details `Input should be a valid UUID, invalid length: expected length 32 for simple format, found 0`

This should work now.

> Also, updating dedupCollId should update autoAddCollections on the backend.

This as well, though need to add tests, but probably ok for feature branch merge.

ikreymer committed:
more update for auto setting autoAddCollections from dedupCollId

ikreymer committed:
further fix for autoAddCollection updating

SuaYoo reviewed (commented):
Tested saving a dedupe collection to a workflow with https://github.com/webrecorder/browsertrix/pull/2893, working well. Haven't tested the other dedupe features yet.

tw4l reviewed (approved):
This now seems to be good! We'll want to add test coverage in the feature branch before merging to main, but I think good to merge to feature branch now to unblock frontend dev
